File: go/test/ecr_api_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/bluefin-tecs-ecr-sdk/go"
	"github.com/voxgig-sdk/bluefin-tecs-ecr-sdk/go/core"

	vs "github.com/voxgig-sdk/bluefin-tecs-ecr-sdk/go/utility/struct"
)

func TestEcrApiEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.EcrApi(nil)
		if ent == nil {
			t.Fatal("expected non-nil EcrApiEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := ecr_apiBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "ecr_api." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BLUEFINTECSECR_TEST_ECR_API_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		ecrApiRef01Ent := client.EcrApi(nil)
		ecrApiRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "ecr_api"}, setup.data), "ecr_api_ref01"))

		ecrApiRef01DataResult, err := ecrApiRef01Ent.Create(ecrApiRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		ecrApiRef01Data = core.ToMapAny(ecrApiRef01DataResult)
		if ecrApiRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LOAD
		ecrApiRef01MatchDt0 := map[string]any{}
		ecrApiRef01DataDt0Loaded, err := ecrApiRef01Ent.Load(ecrApiRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if ecrApiRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func ecr_apiBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "ecr_api", "EcrApiTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read ecr_api test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse ecr_api test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"ecr_api01", "ecr_api02", "ecr_api03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BLUEFINTECSECR_TEST_ECR_API_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BLUEFINTECSECR_TEST_ECR_API_ENTID": idmap,
		"BLUEFINTECSECR_TEST_LIVE":      "FALSE",
		"BLUEFINTECSECR_TEST_EXPLAIN":   "FALSE",
		"BLUEFINTECSECR_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BLUEFINTECSECR_TEST_ECR_API_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["BLUEFINTECSECR_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BLUEFINTECSECR_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBluefinTecsEcrSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BLUEFINTECSECR_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BLUEFINTECSECR_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/EcrApiEntityTest.php
<?php
declare(strict_types=1);

// EcrApi entity test

require_once __DIR__ . '/../bluefintecsecr_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EcrApiEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = BluefinTecsEcrSDK::test(null, null);
        $ent = $testsdk->EcrApi(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = ecr_api_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "ecr_api." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set BLUEFIN_TECS_ECR_TEST_ECR_API_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $ecr_api_ref01_ent = $client->EcrApi(null);
        $ecr_api_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.ecr_api"), "ecr_api_ref01"));

        $ecr_api_ref01_data_result = $ecr_api_ref01_ent->create($ecr_api_ref01_data, null);
        $ecr_api_ref01_data = Helpers::to_map(is_object($ecr_api_ref01_data_result) && method_exists($ecr_api_ref01_data_result, 'data_get') ? $ecr_api_ref01_data_result->data_get() : $ecr_api_ref01_data_result);
        $this->assertNotNull($ecr_api_ref01_data);

        // LOAD
        $ecr_api_ref01_match_dt0 = [];
        $ecr_api_ref01_data_dt0_loaded = $ecr_api_ref01_ent->load($ecr_api_ref01_match_dt0, null);
        $this->assertNotNull($ecr_api_ref01_data_dt0_loaded);

    }
}

function ecr_api_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/ecr_api/EcrApiTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = BluefinTecsEcrSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["ecr_api01", "ecr_api02", "ecr_api03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("BLUEFIN_TECS_ECR_TEST_ECR_API_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "BLUEFIN_TECS_ECR_TEST_ECR_API_ENTID" => $idmap,
        "BLUEFIN_TECS_ECR_TEST_LIVE" => "FALSE",
        "BLUEFIN_TECS_ECR_TEST_EXPLAIN" => "FALSE",
        "BLUEFIN_TECS_ECR_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["BLUEFIN_TECS_ECR_TEST_ECR_API_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["BLUEFIN_TECS_ECR_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["BLUEFIN_TECS_ECR_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new BluefinTecsEcrSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["BLUEFIN_TECS_ECR_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["BLUEFIN_TECS_ECR_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
